### ISSUE TITLE
Some libc tests fail when using clang.

### ISSUE THREAD

**Issue by mikbras** [Microsoft/openenclave#768]
The following **libc** tests fail when using the **clang** compiler (see **tests.broken**).

```
# The following tests fail when built with Clang using these cmake flags.
#     -DCMAKE_BUILD_TYPE=Debug
#     -DENABLE_FULL_LIBC_TESTS=1
../../3rdparty/musl/libc-test/src/functional/tgmath.c
../../3rdparty/musl/libc-test/src/math/fmax.c
../../3rdparty/musl/libc-test/src/math/fmaxf.c
../../3rdparty/musl/libc-test/src/math/fmin.c
../../3rdparty/musl/libc-test/src/math/fminf.c
../../3rdparty/musl/libc-test/src/math/ilogb.c
../../3rdparty/musl/libc-test/src/math/ilogbf.c
../../3rdparty/musl/libc-test/src/math/ilogbl.c
../../3rdparty/musl/libc-test/src/math/pow.c
../../3rdparty/musl/libc-test/src/math/powl.c
../../3rdparty/musl/libc-test/src/math/tgammaf.c
../../3rdparty/musl/libc-test/src/math/y1.c
../../3rdparty/musl/libc-test/src/math/y1f.c
../../3rdparty/musl/libc-test/src/math/yn.c

# The following tests fail when built with Clang using these cmake flags (the above tests also fail).
#     -DCMAKE_BUILD_TYPE=RelWithDebInfo
#     -DENABLE_FULL_LIBC_TESTS=1
../../3rdparty/musl/libc-test/src/math/fdim.c
../../3rdparty/musl/libc-test/src/math/fdimf.c
../../3rdparty/musl/libc-test/src/math/fdiml.c
../../3rdparty/musl/libc-test/src/math/fmaf.c
../../3rdparty/musl/libc-test/src/math/log1p.c
../../3rdparty/musl/libc-test/src/math/log1pf.c
../../3rdparty/musl/libc-test/src/math/powf.c
```

**Notes:**
- The second group of failures can be worked around by suppressing Clang optimizations.
- Compiling **MUSL** with GCC and the tests with Clang works.
- The failures could be due to pre-compiler differences (try the -E option and compare).
- the fmax and fmin failures are caused by Clang generated builtin versions of these.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/fmax.c
../../3rdparty/musl/libc-test/src/math/fmaxf.c
../../3rdparty/musl/libc-test/src/math/fmin.c
../../3rdparty/musl/libc-test/src/math/fminf.c

For fmax, fmaxf, fmin, fminf function, clang emits the operations in place instead of calling our library function. Therefore the behavior is different. The failing cases are when one of the arguments is nan or inf. Our library function has special case for nans.
Nice to do: Investigate why inf case fails.
This is easily reproducible in a standalone C program that calls fmax or fmin by compiling under GCC, clang and passing in nan, inf for one of the parameters.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/ilogb.c
../../3rdparty/musl/libc-test/src/math/ilogbf.c
../../3rdparty/musl/libc-test/src/math/ilogbl.c

These fail for 0, inf inputs. The reason is that the musl implementation is using
FORCE_EVAL(0/0.0f) 
to trigger setting the exception flags.
Clang optimizes away the divide instruction even in debug mode and replaces it with nan.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/pow.c

For large values, MUSL implementation return huge*huge where huge=1.0e+300. huge*huge should set overflow flag. But clang evaluates it at compile time to Inf.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/powl.c

Clang ends up compile-time evaluating the expression 1.0/0.0 in musl implementation.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/tgammaf.c

Clang ends up compile-time evaluating the expression 0/0.0 in musl implementation.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/y1.c
../../3rdparty/musl/libc-test/src/math/y1f.c
../../3rdparty/musl/libc-test/src/math/yn.c

Clang ends up compile-time evaluating the expressions 0/0.0

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/functional/tgmath.c I get linker error.:
undefined reference to `csqrt'

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/fmaf.c
passed in RelWithDebInfo mode.

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/log1p.c
../../3rdparty/musl/libc-test/src/math/log1pf.c

Clang ends up compile-time evaluating the expressions x/0.0, (x-x)/0.0

**Comment by anakrish** [comment]
../../3rdparty/musl/libc-test/src/math/powf.c

There is code in MUSL powf :
 if (ix < 0x00800000) {  
      ax *= two24;  
      n -= 24;     
     GET_FLOAT_WORD(ix, ax);
}
Clang replaces this with conditional move where ax * two24 is always executed, changing floating-point exception flags.

**Comment by anakrish** [comment]
@CodeMonkeyLeet I have provided that analysis for all the tests in the list. 
It is due to clang's compile-time evaluation of expressions with floating-point side effects, and also possible due to a bug where it replaces if-else with a conditional move.
I recommend closing this issue.

**Comment by anitagov** [comment]
@anakrish - We should move this to tests.unsupported in that case. Also file/point to a clang bug if one exists.

**Comment by anitagov** [comment]
@anakrish - We should remove the tests from tests.broken and modify CMake to build this only for gcc and skip for clang. Also file/point to a clang bug if one exists.  Thanks.

**Comment by CodeMonkeyLeet** [comment]
@berinpaul can you take these tests in tests.broken and move them into test.support, then update the tests/libc/CMakeLists.txt so that they only get run on the GCC configuration?

**Comment by anitagov** [comment]
@berinpaul - Hasn't this been completed?

**Comment by berinpaul** [comment]
Yes. Issue #1090 opened to track the remaining test in tests.broken.

**Comment by CodeMonkeyLeet** [comment]
Resolved by #1049
